context/src/main/java/io/grpc/Context.java

```diff
@@ -109,37 +109,42 @@
    */
   public static final Context ROOT = new Context(null);
 
-  // One and only one of them is non-null
-  private static final Storage storage;
-  private static final LinkedBlockingQueue<ClassNotFoundException> storageOverrideNotFoundErrors =
-      new LinkedBlockingQueue<ClassNotFoundException>();
-  private static final Exception storageInitError;
+  // Lazy-loaded storage. Delaying storage initialization until after class initialization makes it
+  // much easier to avoid circular loading since there can still be references to Context as long as
+  // they don't depend on storage, like key() and currentContextExecutor(). It also makes it easier
+  // to handle exceptions.
+  private static final AtomicReference<Storage> storage = new AtomicReference<Storage>();
 
-  static {
-    Storage newStorage = null;
-    Exception error = null;
+  // For testing
+  static Storage storage() {
+    Storage tmp = storage.get();
+    if (tmp == null) {
+      tmp = createStorage();
+    }
+    return tmp;
+  }
+
+  private static Storage createStorage() {
+    // Note that this method may be run more than once
     try {
       Class<?> clazz = Class.forName("io.grpc.override.ContextStorageOverride");
```

Comment:
@raphw I always wondered if there's a bytebuddy way to test code like this. For example, a test class which declares a good and various naughty impls of this type. That way, we could test these defenses.